include/nbl/video/alloc/SubAllocatedDescriptorSet.h

```diff
@@ -0,0 +1,119 @@
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef _NBL_VIDEO_SUB_ALLOCATED_DESCRIPTOR_SET_H_
+#define _NBL_VIDEO_SUB_ALLOCATED_DESCRIPTOR_SET_H
+
+#include "nbl/video/alloc/IBufferAllocator.h"
+
+#include <type_traits>
+
+namespace nbl::video
+{
+
+// address allocator gives offsets
+// reserved allocator allocates memory to keep the address allocator state inside
+template<class AddrAllocator, class ReservAllocator=core::allocator<uint8_t>>
+class SubAllocatedDescriptorSet : public core::IReferenceCounted 
+{
+  public:
+        using AddressAllocator = AddrAllocator;
+        using ReservedAllocator = ReservAllocator;
+        using size_type = typename AddressAllocator::size_type;
+        using value_type = typename AddressAllocator::size_type;
+        static constexpr value_type invalid_value = AddressAllocator::invalid_address;
+
+        // constructors
+        template<typename... Args>
+        inline SubAllocatedDescriptorSet(const std::span<const video::IGPUDescriptorSetLayout::SBinding> bindings, 
+            ReservedAllocator&& _reservedAllocator, const value_type maxAllocatableAlignment, Args&&... args)
+        {
+            auto allocatableDescriptors = 0;
+            m_allocatableRanges.reserve(bindings.size());
+
+            for (auto& binding : bindings)
+            {
+                SubAllocDescriptorSetRange range;
+                range.offset = allocatableDescriptors;
+                range.binding = binding;
+                // Only bindings with these flags will be allocatable
+                if (binding.createFlags.hasFlags(core::bitflag(IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_UPDATE_AFTER_BIND_BIT) 
+					| IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_UPDATE_UNUSED_WHILE_PENDING_BIT 
+					| IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_PARTIALLY_BOUND_BIT))
+                {
+                    allocatableDescriptors += binding.count;
+                }
+				m_allocatableRanges.push_back(range);
+            }
+
+			m_addressAllocator = AddrAllocator(
+				_reservedAllocator.allocate(AddressAllocator::reserved_size(maxAllocatableAlignment, static_cast<size_type>(allocatableDescriptors), args...), _NBL_SIMD_ALIGNMENT),
+				static_cast<size_type>(0), 0u, maxAllocatableAlignment, static_cast<size_type>(allocatableDescriptors), std::forward<Args>(args)...
+			);
+			m_reservedAllocator = ReservedAllocator(std::move(_reservedAllocator));
+            m_reservedSize = allocatableDescriptors;
```

Comment:
nooo, you need to prime it with `AddressAllocator::reserved_size` ou allocated the reserved space with